controller/controller_server/csi_controller_server.py

```diff
@@ -388,10 +388,37 @@ def ControllerUnpublishVolume(self, request, context):
             return csi_pb2.ControllerUnpublishVolumeResponse()
 
     def ValidateVolumeCapabilities(self, request, context):
+        set_current_thread_name(request.volume_id)
         logger.info("ValidateVolumeCapabilities")
-        context.set_code(grpc.StatusCode.UNIMPLEMENTED)
-        logger.info("finished ValidateVolumeCapabilities")
-        return csi_pb2.ValidateVolumeCapabilitiesResponse()
+        try:
+
+            utils.validate_validate_volume_capabilities_request(request)
+
+            secrets = request.secrets
+            user, password, array_addresses = utils.get_array_connection_info_from_secret(secrets)
+
+            # TODO : pass multiple array addresses
+            array_type, volume_id = utils.get_volume_id_info(request.volume_id)
+
+            with get_agent(user, password, array_addresses, array_type).get_mediator() as array_mediator:
+                volume = array_mediator.get_volume_by_id(volume_id=volume_id)
+
+            if request.volume_context:
+                utils.validate_volume_context_match_volume(request.volume_context, volume)
+            logger.info("finished ValidateVolumeCapabilities")
+            return utils.generate_csi_validate_volume_capabilities_response(request.volume_context,
+                                                                            request.volume_capabilities,
+                                                                            request.parameters)
+        except (controller_errors.VolumeNotFoundError, ObjectIdError) as ex:
+            logger.exception(ex)
+            context.set_details(ex.message)
+            context.set_code(grpc.StatusCode.NOT_FOUND)
+            return csi_pb2.ValidateVolumeCapabilitiesResponse()
```

Comment:
consider adding `ex.message` to `ValidateVolumeCapabilitiesResponse` as well

Reply:
i dont think it will help us or even possible..

Reply:
why not?

Reply:
oh, we have a message value in response, I will add the msg

Reply:
done